src/Functions/FunctionStringCompare.cpp

```diff
@@ -0,0 +1,414 @@
+    {
+        chassert(arguments[0].column->size() == input_rows_count);
+        auto col_result = ColumnInt8::create();
+        if (!input_rows_count)
+            return std::move(col_result);
+
+        auto prepare_string_args
+            = [](const ColumnPtr & column) -> std::tuple<const ColumnString *, const ColumnFixedString *, const ColumnConst *>
+        {
+            return {
+                checkAndGetColumn<ColumnString>(column.get()),
+                checkAndGetColumn<ColumnFixedString>(column.get()),
+                checkAndGetColumn<ColumnConst>(column.get())};
+        };
+        auto [col_str1, fixed_str1_col, col_str1_const] = prepare_string_args(arguments[0].column);
+        auto [col_str2, fixed_str2_col, col_str2_const] = prepare_string_args(arguments[1].column);
+
+        auto str1_offset = arguments.size() > 2 ? arguments[2].column->getUInt(0) : 0;
+        auto str2_offset = arguments.size() > 2 ? arguments[3].column->getUInt(0) : 0;
+
+        auto num_bytes = arguments.size() > 2 ? arguments[4].column->getUInt(0) : std::numeric_limits<size_t>::max();
+        if (num_bytes == 0)
+        {
+            col_result->insertMany(0, arguments[0].column->size());
+            return std::move(col_result);
+        }
+
+        col_result->getData().resize(arguments[0].column->size());
+        auto & result_array = col_result->getData();
+
+        if (col_str1 && col_str2_const)
+            executeStringConst<false>(*col_str1, str1_offset, String(col_str2_const->getDataAt(0)), str2_offset, num_bytes, result_array);
+        else if (col_str1_const && col_str2)
+            executeStringConst<true>(*col_str2, str2_offset, String(col_str1_const->getDataAt(0)), str1_offset, num_bytes, result_array);
+        else if (col_str1 && col_str2)
+            executeStringString(*col_str1, str1_offset, *col_str2, str2_offset, num_bytes, result_array);
+        else if (fixed_str1_col && fixed_str2_col)
+            executeFixedStringFixedString(*fixed_str1_col, str1_offset, *fixed_str2_col, str2_offset, num_bytes, result_array);
+        else if (fixed_str1_col && col_str2)
+            executeFixedStringString<false>(*fixed_str1_col, str1_offset, *col_str2, str2_offset, num_bytes, result_array);
+        else if (col_str1 && fixed_str2_col)
+            executeFixedStringString<true>(*fixed_str2_col, str2_offset, *col_str1, str1_offset, num_bytes, result_array);
+        else if (fixed_str1_col && col_str2_const)
+            executeFixedStringConst<false>(*fixed_str1_col, str1_offset, String(col_str2_const->getDataAt(0)), str2_offset, num_bytes, result_array);
+        else if (fixed_str2_col && col_str1_const)
+            executeFixedStringConst<true>(*fixed_str2_col, str2_offset, String(col_str1_const->getDataAt(0)), str1_offset, num_bytes, result_array);
+        else if (col_str1_const && col_str2_const)
```

Comment:
The const/const case is in practice not relevant. If you add

```cpp
bool useDefaultImplementationForConstants() const override { return true; }
```

to l. 41, then const/const argument will be internally converted to non-const/non-const arguments and we can delete the implementation of `executeConstConst` entirely.